keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,964 @@
+     within those layers. The artifact manifest can have media types that differ from those in standard container images.
+   - **Use Case:** Allows the distribution of non-container content using the same infrastructure and tools developed for OCI images.
+
+**OCI Object:**
+   - Umbrella term encompassing both OCI images and OCI artifacts. It represents any object that conforms to the OCI specifications for storage and distribution.
+
+### Risks and Mitigations
+
+- **Security Risks:** Allowing direct mounting of OCI images introduces potential attack vectors. Mitigation includes thorough security reviews and
+  limiting access to trusted registries. Limiting to OCI artifacts (non-runnable content) or read-only mode may lessen the security risk.
+- **Compatibility Risks:** Existing webhooks watching for the images used by the pod using some policies will need to be updated to expect the image to be specified as a `VolumeSource`.
+- **Performance Risks:** Large images or artifacts could impact performance. Mitigation includes optimizations in the implementation and providing
+  guidance on best practices for users.
+
+## Design Details
+
+The new `VolumeSource` will be defined in the Kubernetes API, and the implementation will involve updating components (CRI, Kubelet)
+to support this source type. Key design aspects include:
+
+- API changes to introduce the new `VolumeSource` type.
+- Modifications to the Kubelet to handle mounting OCI images and artifacts.
+- Handling image pull secrets and registry authentication.
+- Reuse existing logic from ConfigMaps for:
+  - Determining the file location on the host filesystem.
+  - Managing garbage collection of unused images and artifacts.
+  - Handling updates to the OCI image or artifact, similar to how ConfigMaps update mounted files.
+
+The following code snippet illustrates the proposed API change:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: example-pod
+spec:
+  volumes:
+  - name: oci-volume
+    oci:
+      image: "example.com/my-image:latest"
+  containers:
+  - name: my-container
+    image: busybox
+    volumeMounts:
+    - mountPath: /data
+      name: oci-volume
+```
+
+### Kubelet and Container Runtime Interface (CRI) support for OCI artifacts
+
+Kubelet and the Container Runtime Interface (CRI) currently handle OCI images. To support OCI artifacts,
+potential enhancements may be required:
+
+**Extended Media Type Handling:**
+   - Update Kubelet and CRI to recognize and handle new media types associated with OCI artifacts.
+   - Ensure that pulling and storing these artifacts is as efficient and secure as with OCI images.
+
+**Lifecycling and Garbage Collection:**
+   - Implement custom logic for managing the lifecycle of various OCI artifacts.
+
+**Artifact-Specific Configuration:**
+   - Introduce new configuration options or CRD to handle the unique requirements of different types of OCI artifacts.
```

Comment:
how CRDs will work?
this raises another question, is this option available for disconnected kubelets?

Reply:
> disconnected kubelets

Do you mean “can static Pods use this volume source”?

Reply:
I feel like they should be treated the same as hostPath (I'm not sure if those work with static pods off the top of my head)

Reply:
Static Pods **can** use `hostPath` (it's quite common). IIRC they can't use `imagePullSecrets` and a few things like that, though.

Reply:
I removed the note to the CRD, I think it's a bit out of scope for now.